Migrate to zod

FuzzB0t opened the pull request:
WIP - Ref #270 

Migrated all instances of AJV validation to zod, this PR is still a work in progress as there is more testing and refining to be done.

Here's what's done so far:
- [x] Changed all JTD schemas into zod schemas
- [x] Removed AJV library from repo
- [x] Successfully ran test suite + manual validation testing
- [x] Added docgen for schema types
- [x] Added cannonfile spec view to website

An example of manual validation testing is the following:

Validation Error on wrong required keyname "artifacts":
![image](https://github.com/usecannon/cannon/assets/30664234/6648633d-062a-4e11-b607-6de587320246)


Validation error on invalid required key type:
![image](https://github.com/usecannon/cannon/assets/30664234/2097816b-5f9f-4dba-beca-f6250aba85a3)

Successful build after fixing all errors:
![image](https://github.com/usecannon/cannon/assets/30664234/6e073489-360d-47b6-84b3-2137b9b2080f)

dbeal-eth commented on packages/builder/src/actions.ts:
```diff
@@ -46,29 +45,50 @@ export const ActionKinds: { [label: string]: CannonAction } = {};
+        description: z.string(),
+        keywords: z.array(
+          z.string({
+            invalid_type_error: 'keywords must be strings',
```
Keywords must be an array of strings

FuzzB0t replied:
Yeah it is an array of strings, I'll change the error message to better resemble that but this error pops up if any of the keywords inside the array aren't strings.

dbeal-eth commented on packages/builder/src/steps/invoke.ts:
```diff
@@ -107,7 +151,7 @@ async function runTxn(
   }
 
-  if (config.fromCall && !contract.functions[config.fromCall.func]) {
+  if (config.fromCall && config.fromCall.func && !contract.functions[config.fromCall.func]) {
```
Did you run into an issue with this? Seems like unrelated change

FuzzB0t replied:
I did, It was potentially coming back undefined since the fromCall parameter is optional, zod handles optional parameters as a union of ```type | undefined```

dbeal-eth commented on packages/builder/src/types.zod.ts:
```diff
@@ -0,0 +1,280 @@
+import { ethers } from 'ethers';
```
This is really cool, having all the internal storage types in zod as well. Does this mean we will have a way to validate ipfs artifacts just to make sure? Can we use it as a way to verify that updates to cannon schema are non-breaking progmatically

FuzzB0t replied:
Yes! That's the idea, it's not fully implemented yet but the idea is that we can use zod as our source of truth which provides us inherent validation for our core types while being able to generate types directly from our zod schemas.

dbeal-eth commented on packages/builder/src/util.ts:
```diff
@@ -4,21 +4,38 @@ import { Buffer } from 'buffer';
-    env: { elements: { type: 'string' } },
-  },
-} as const;
+// FIXME: Why not move this to keepers.ts since this isnt shared by any other file?
```
This used to be used be used for both keepers and `run` step (which was moved into cli obviously)

So I would agree it makes sense to make this change

dbeal-eth reviewed (commented):
Looking really good! Not a fine tooth comb reciew but would like to see much better error lofting format (than dumping a Json object). Do you have any ideas for that?

FuzzB0t commented:
> Looking really good! Not a fine tooth comb reciew but would like to see much better error lofting format (than dumping a Json object). Do you have any ideas for that?

Yes, so right now im using zod's ```.parse``` method which throws a ZodError object (the JSON dump). I can use ```safeParse``` to silently throw an error and catch these errors and create custom error handlers like in [this part](https://zod.dev/?id=error-formatting) of the zod documentation.

FuzzB0t commented:
Remaining tasks list for progress reference:

- [ ] Improve error logging
- [ ] Add docgen from zod types (using something like [this](https://github.com/samchungy/zod-openapi))
- [ ] Look into adding type inference from zod types to ts (This is to keep zod as our source of truth and be able to validate results or inputs from things like IPFS or JSON-RPC calls)
- [ ] Add guidelines to builder README for adding new input types

FuzzB0t commented:
Fixing errors, they now look like this:

![image](https://github.com/usecannon/cannon/assets/30664234/491f946e-70dd-4552-9938-bfe53ff81ab1)

This format is easily customizable and can be changed to match another kind of pattern.

Also added a way to autogenerate markdown docs from the ts inferred zod schemas.

FuzzB0t commented:
Added html docgen  with Typedoc:

![image](https://github.com/usecannon/cannon/assets/30664234/a3b1abec-235f-4f9c-80c3-2d6737cc7fac)

Example Step Config Documentation (HTML):
![image](https://github.com/usecannon/cannon/assets/30664234/9ad86397-5e03-4b88-9515-b86773c7bc12)

Also added Markdown Docgen (for github):
![image](https://github.com/usecannon/cannon/assets/30664234/85cbab46-5fd6-46a5-ba49-c7dc6c19cf3b)

FuzzB0t commented on packages/builder/src/schemas.zod.ts:
Contains all of the step config schemas that were validated with AJV

FuzzB0t commented on packages/builder/src/schemas.zod.ts:
```diff
@@ -0,0 +1,297 @@
+
+// All properties are made optional since all schemas that
+// are validated against this are a partial object of this schema.
+export type ConfigValidationSchema = z.ZodObject<{
```
General validation schema to use to validate step configs against

FuzzB0t commented on packages/website/content/docs-technical.md:
```diff
@@ -68,7 +68,6 @@ Publish a Cannon package to the registry for all networks where this package has
 **Options**
 
 - `--chain-id` - The network which the registry to deploy to is on. Default: 1 (mainnet)
-- `--provider-url` - RPC endpoint to use for publishing. If not specified, uses a local wallet provider (such as Frame)
```
This piece in the docs was incorrect, after looking through the publish.ts file there was no reference to use of the "provider-url" param

FuzzB0t commented:
Successfully built the synthetix-sandbox package:

```
n00b@DESKTOP-C38BT9L:~/work/cannon-packages/synthetix-sandbox$ cannon-dev build 
Building the foundry project using forge build...
forge build succeeded
Checking IPFS for package synthetix-sandbox:3.1.2...
Existing package found.

Using package...
Name: synthetix-sandbox
Version: 3.1.2
Preset: main (default)
Source code will be included in the package

Building the chain (ID 13370 via frame) into the state defined in cannonfile.toml...
plugins: cannon-plugin-router detected
exec: provision.synthetix
Checking IPFS for package synthetix:latest...
  exec: contract.AccountModule
  exec: contract.AssociateDebtModule
  exec: contract.CcipReceiverModule
  exec: contract.CollateralConfigurationModule
  exec: contract.CollateralModule
  exec: contract.CrossChainUSDModule
  exec: contract.IssueUSDModule
  exec: contract.LiquidationModule
  exec: contract.MarketCollateralModule
  exec: contract.MarketManagerModule
  exec: contract.MulticallModule
  exec: contract.PoolConfigurationModule
  exec: contract.PoolModule
  exec: contract.RewardsManagerModule
  exec: contract.USDTokenModule
  exec: contract.UtilsModule
  exec: contract.VaultModule
  exec: contract.InitialCoreProxy
  exec: router.AccountRouter
  exec: router.CoreRouter
  exec: router.USDRouter
  exec: invoke.upgrade_core_proxy
  exec: provision.oracle_manager
Checking IPFS for package oracle-manager:3.2.1-dev.f440e7fa...
    exec: contract.NodeModule
    exec: contract.InitialProxy
    exec: router.OracleRouter
    exec: invoke.upgrade_proxy
  exec: invoke.enable_basic_features
  exec: invoke.init_account
  exec: invoke.init_usd
  exec: invoke.register_const_one_oracle
  exec: invoke.set_oracle_manager
  exec: invoke.configure_usd_collateral
exec: invoke.allowAllCreatePool
exec: invoke.allowAllRegisterMarket
exec: invoke.createAccount
exec: invoke.registerCollateralPriceNode
exec: invoke.setMinimumLiquidityRatio
exec: provision.spotMarket
Checking IPFS for package synthetix-spot-market:latest...
  exec: import.synthetix
Checking IPFS for package synthetix:latest...
  exec: contract.AsyncOrderConfigurationModule
  exec: contract.AsyncOrderModule
  exec: contract.AsyncOrderSettlementModule
  exec: contract.AtomicOrderModule
  exec: contract.MarketConfigurationModule
  exec: contract.SpotMarketFactoryModule
  exec: contract.WrapperModule
  exec: contract.InitialSpotMarketProxy
  exec: router.SpotMarketRouter
  exec: router.SynthRouter
  exec: invoke.upgradeSpotMarketProxy
  exec: invoke.addCreateSynthToFeatureFlag
  exec: invoke.setSynthImplementation
  exec: invoke.setSynthetixSystem
  exec: invoke.addSpotMarketToFeatureFlag
exec: invoke.configureTknCollateral
exec: invoke.createPool
exec: invoke.createSynth
exec: invoke.mintCollateral
exec: provision.perpsMarket
Checking IPFS for package synthetix-perps-market:latest...
  exec: import.spotMarket
Checking IPFS for package synthetix-spot-market:latest...
  exec: contract.AsyncOrderModule
  exec: contract.AsyncOrderSettlementModule
  exec: contract.CoreModule
  exec: contract.FeatureFlagModule
  exec: contract.GlobalPerpsMarketModule
  exec: contract.LiquidationModule
  exec: contract.MarketConfigurationModule
  exec: contract.PerpsAccountModule
  exec: contract.PerpsMarketFactoryModule
  exec: contract.PerpsMarketModule
  exec: import.synthetix
Checking IPFS for package synthetix:latest...
  exec: contract.InitialProxy
  exec: router.PerpsMarketRouter
  exec: invoke.upgrade_proxy
  exec: invoke.addPerpsMarketToFeatureFlag
  exec: invoke.init_account
  exec: invoke.setSpotMarket
  exec: invoke.setSynthetix
  exec: invoke.addCreateMarketToFeatureFlag
exec: invoke.addAtomicFixedFee
exec: invoke.allowAllCreatePerpsMarket
exec: invoke.allowCreatePerpsAccount
exec: invoke.approveDeposit
exec: invoke.setMarketCollateralSupplyCap
exec: invoke.setMaxCollateralAmount
exec: invoke.setPreferredPool
exec: invoke.updateSynthPriceData
exec: invoke.createPerpsMarket
exec: invoke.depositCollateral
exec: invoke.initializeWrapper
exec: invoke.setPoolConfiguration
exec: invoke.delegateCollateral
exec: invoke.updatePriceFeed
exec: invoke.mintStablecoins

TRANSACTIONS:
╔══════════════════════════════╤════════════════════════════════════════════════════════════════════╗
║ createSynth                  │ 0xfca7673c34a2fd00bbd42201e1fd820dd9b1993f7a7374040b1492a27822a227 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ addAtomicFixedFee            │ 0xe916b2348a7cd84c4ca1d48a122bc55c9c0f3dd472460ff91452122f0b80adb6 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ allowAllCreatePerpsMarket    │ 0x2853f03700b95eb81852c15b57185f4c7ac87d9dafb29cf959b679da49b7a7a6 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ allowAllRegisterMarket       │ 0x498994487073749c1699da7f28f687a0bdb2382bbed5f38e711cbf846d984125 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ registerCollateralPriceNode  │ 0x8813a4e399110888ffe9b3cfd3e99b104dc4a8038471809f44c8e453dcd6f946 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ configureTknCollateral       │ 0xc95609edbab07ac7b79042d932c56dc946b7948f2121464824a4e94139a76dfc ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ setMarketCollateralSupplyCap │ 0xd2ee436ae3f2f679ed1f3c798aa1792442f88bfa91250b1999c9188c9fdfe654 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ initializeWrapper            │ 0x3afe4e28d43a19d6ded3278cd29e47b51530172ebd71297fd933d0e457233da3 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ createAccount                │ 0x93b03111e547311073e2423f4d5793ba687ff7923b65ed2b6e5d024c9f22c8ce ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ mintCollateral               │ 0x7e4a2a9089c99a454c2bf7602c186abf1dacf97753e25d5b275cd4b9244ac130 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ approveDeposit               │ 0x8f4227e9d47514c0b74935dee2430d1f2725a1c267e7a4b9665cc42e26022c59 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ depositCollateral            │ 0x7b75af7c2257a1cbbbcd9b85dffb862d34c2b7731ef3a68541008c63309e0643 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ delegateCollateral           │ 0x2fdb4b039c65fa5e0bb6a0068edabdbb2bcc86e95c23d8f47f03da7c4942f31f ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ mintStablecoins              │ 0x6aba9c256b7c52b46f747ccafc0acd1e38ceb094872f3d522614d8ed9b1da667 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ setMaxCollateralAmount       │ 0x6c315f955f94b187832a6824b29c6fa78792c551fb9667fc1fa7b1f5236994aa ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ setMinimumLiquidityRatio     │ 0xdfb4cf8ed0ccb6037a07dbb21fa709faf6c88bac496eb555ceb200ed70cd5caa ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ allowAllCreatePool           │ 0x65197791e11b89bd07fa8b00428772a0fd60c1ec1156d0049c6856af526fc11f ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ createPool                   │ 0xfe023123e4c569af8c64377099e31ad407ae60e19442e3ba3c1c5292bee786ca ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ updateSynthPriceData         │ 0x199682473c9bbaa1306c701e4cbcf8398c368487a08040919e77d25078090a0e ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ setPoolConfiguration         │ 0xcfc75dfc52aadcea94b2991a13dc2cd2dd1702227ddce9031796bc7f370dff5e ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ setPreferredPool             │ 0x77d8d835614f6b864c3cc8fe83fb7d9f6f3b822c2ebfc725714b90498cf73add ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ allowCreatePerpsAccount      │ 0x7ae1afbcde8f46c3953a65ccf772cb3a6d2a0ed28c5762cbb620ed41b4ca1e60 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ createPerpsMarket            │ 0xd44ef0e649f20eb8ea5050a454b4fec9b47bb04f17415cd10b3f2c0a8d478a22 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ updatePriceFeed              │ 0x73b6706868f85fcfbe2c7ac81382a00671477e830e022f59dd2739fbffea2484 ║
╚══════════════════════════════╧════════════════════════════════════════════════════════════════════╝


EXTRA DATA:
╔═══════════════╤════════════════════════════════════════════════════════════════════╗
║ synthMarketId │ 1                                                                  ║
╟───────────────┼────────────────────────────────────────────────────────────────────╢
║ tkn_oracle_id │ 0x6cb1042857d7329f5ee7ae88c96fd693a39b006b6a8884139c99802c856ca588 ║
╟───────────────┼────────────────────────────────────────────────────────────────────╢
║ perpsMarketId │ 2                                                                  ║
╚═══════════════╧════════════════════════════════════════════════════════════════════╝

Successfully built package synthetix-sandbox:3.1.2 (ipfs://QmcZRdePJHgT5Xo2V1HWj8P9aCfWJWyoALJncCEzSFk8K5)

```

dbeal-eth commented on packages/builder/src/actions.ts:
```diff
@@ -1,14 +1,15 @@
-const ajv = new Ajv();
-
 import contractSpec from './steps/contract';
+
```
random extra spaces being added here?

FuzzB0t replied:
Yep good catch, just removed these.

dbeal-eth commented on packages/builder/src/actions.ts:
```diff
@@ -82,11 +58,15 @@ export function registerAction(action: CannonAction) {
+      keywords: true,
+      setting: true,
+      import: true,
+    }) as any
```
what is the purpose of using `.pick` and then `as any` afterwards here?

FuzzB0t replied:
Was trying to match the way it was being done before, also, as any was the only way I could index the object returned by `pick` with the `label` string

dbeal-eth commented on packages/builder/src/definition.ts:
```diff
@@ -25,15 +25,18 @@ export type StateLayers = {
   readonly allActionNames: string[];
 
-  // actions which have no dependencies
+  /** actions which have no dependencies */
```
is jekyll not compatible with `//` and you have to use `/** */`?

FuzzB0t replied:
No this is to enable TSDOC comments which help with documenting certain things. Also have some useful features like being able to see these comments on instances of this class.

dbeal-eth commented on packages/builder/src/schemas.zod.ts:
```diff
@@ -0,0 +1,304 @@
+
+/// ================================ INPUT CONFIG SCHEMAS ================================ \\\
+
+export const contractSchema = z
```
so for the builder base cannon steps its all well to put the zod definitions in this file. However, what happens for custom steps for validation (ex. `cannon-plugin-router` https://github.com/Synthetixio/cannon-plugin-router )

I *personally* am not a big fan of putting all the types for a package into a single file like some sort of a dumping ground

FuzzB0t replied:
Hmmm thats a good point. Existing plugins will have to switch their schemas to zod. However its worth rethinking a few things based on this:

- Before migrating, we validated any incoming "action" config schema as a JSON schema with properties and optional properties, any parameter was valid as long as it followed basic JTDDataType guidelines.          
- I realized that the ConfigValidationSchema that is used to validate any action schema the way I wrote it is probably not gonna work, since we need to sort of be able to do a "catchall" on config properties created for custom plugins.

My suggestion for that bug is to change the ConfigValidationSchema to be way less strict and see if there's a way for us to allow for new parameters in custom actions.

As for putting everything in one file: I don't personally like it either, it just makes generating the docs a bit easier but i can achieve the same output with a little extra work if we prefer to have the types in their respective action file.

dbeal-eth commented on packages/builder/src/schemas.zod.ts:
```diff
@@ -0,0 +1,304 @@
+  .merge(
+    z
+      .object({
+        create2: z.boolean(),
```
can we share an example of how we would add documentation description text to these types? I would like to do so myself if we think its out of scope, but it would be great to be able to just encode descriptions for what all of these fields do right in the code!

FuzzB0t replied:
Yes will add this into my next batch of changes.

FuzzB0t replied:
Here's a guide on how to do this with typedoc: 

https://typedoc.org/guides/doccomments/

dbeal-eth commented on packages/builder/src/schemas.zod.ts:
```diff
@@ -0,0 +1,304 @@
+        nonce: z.string(),
+        abi: z.string(),
+        abiOf: z.array(z.string()),
+        args: z.array(z.any()),
```
might actually be possible to constrain this type a little bit going forward. for example the type could be like `type argsType = string | number | argsType[] | Record<string, argsType>`

dbeal-eth replied:
i know this is not how the previous ajv definition worked but looking at minor improvements

dbeal-eth commented on packages/builder/src/schemas.zod.ts:
```diff
@@ -0,0 +1,304 @@
+    z
+      .object({
+        create2: z.boolean(),
+        from: z.string(),
```
could we have a zod custom validation for an address? the validation would just be `ethers.utils.isAddress(f)`. Definitely something that can be verified as part of syntax

dbeal-eth commented on packages/builder/src/schemas.zod.ts:
```diff
@@ -0,0 +1,304 @@
+      .object({
+        create2: z.boolean(),
+        from: z.string(),
+        nonce: z.string(),
```
even though the type is a string, it also must be numeric (either 0x prefixed hex or regular decimal number)

dbeal-eth commented on packages/builder/src/schemas.zod.ts:
```diff
@@ -0,0 +1,304 @@
+        create2: z.boolean(),
+        from: z.string(),
+        nonce: z.string(),
+        abi: z.string(),
```
should be JSON parsable even though it is a string

dbeal-eth commented on packages/builder/src/schemas.zod.ts:
```diff
@@ -0,0 +1,304 @@
+
+        value: z.string(),
+        overrides: z.object({
+          gasLimit: z.string(),
```
same as nonce this has to be numeric

dbeal-eth commented on packages/builder/src/schemas.zod.ts:
```diff
@@ -0,0 +1,304 @@
+          gasLimit: z.string(),
+        }),
+
+        depends: z.array(z.string()),
```
this shuold be validated to be `<step type>.<name>` format. can probably do regex match here at least.

dbeal-eth commented on packages/builder/src/schemas.zod.ts:
```diff
@@ -0,0 +1,304 @@
+
+export const invokeSchema = z
+  .object({
+    target: z.array(z.string()).nonempty(),
```
validate that its either an address or valid cannon step name (cannon step names can only have certain characters I believe?)

FuzzB0t replied:
I can validate that its either an address or interpolated value, if we start to validate "known" step names it gets tricky since new plugins can add new step names

dbeal-eth commented on packages/builder/src/schemas.zod.ts:
```diff
@@ -0,0 +1,304 @@
+  .merge(
+    z
+      .object({
+        abi: z.string(),
```
shoudl be json parsable

dbeal-eth commented on packages/builder/src/schemas.zod.ts:
```diff
@@ -0,0 +1,304 @@
+      .object({
+        abi: z.string(),
+
+        args: z.array(z.any()),
```
coudl be a bit more constrained than any

dbeal-eth commented on packages/builder/src/schemas.zod.ts:
```diff
@@ -0,0 +1,304 @@
+        abi: z.string(),
+
+        args: z.array(z.any()),
+        from: z.string(),
```
must be address

dbeal-eth replied:
could also be interpolation

dbeal-eth commented on packages/builder/src/schemas.zod.ts:
```diff
@@ -0,0 +1,304 @@
+              args: z.array(z.any()),
+            })
+          ),
+        value: z.string(),
```
must be numeric. could be decimal (ie `1.234`)

dbeal-eth commented on packages/builder/src/schemas.zod.ts:
```diff
@@ -0,0 +1,304 @@
+        from: z.string(),
+        fromCall: z
+          .object({
+            func: z.string(),
```
the usage of `.deepPartial` on line 94 will cause `func` to be optional here if I am not mistaken, but the field is not optional if `fromCall` is specified

dbeal-eth replied:
https://github.com/colinhacks/zod#deeppartial

dbeal-eth commented on packages/builder/src/schemas.zod.ts:
```diff
@@ -0,0 +1,304 @@
+
+export const provisionSchema = z
+  .object({
+    source: z.string(),
```
validate this has cannon package format (or interpolation) `packageName:packageVersion`

dbeal-eth commented on packages/builder/src/schemas.zod.ts:
```diff
@@ -0,0 +1,304 @@
+ */
+export const chainDefinitionSchema = z
+  .object({
+    name: z.string(),
```
name must not be greater than 31 characters and must have no special characters https://github.com/usecannon/cannon/blob/main/packages/registry/contracts/CannonRegistry.sol#L31

dbeal-eth commented on packages/builder/src/schemas.zod.ts:
```diff
@@ -0,0 +1,304 @@
+export const chainDefinitionSchema = z
+  .object({
+    name: z.string(),
+    version: z.string(),
```
version must not be longer than 31 characters and may have other restrictions

dbeal-eth commented on packages/builder/src/schemas.zod.ts:
```diff
@@ -0,0 +1,304 @@
+          z
+            .object({
+              description: z.string(),
+              type: z.enum(['number', 'string', 'boolean']),
```
type probably isnt needed at this point, we havent ever used it tmk. do you think it still has a use?

FuzzB0t replied:
Havent really seen it used anywhere, so it's probably ok to remove. Will double check to make sure there isnt any package that happens to be using it so we avoid introducing breaking changes.

dbeal-eth commented on packages/builder/src/schemas.zod.ts:
```diff
@@ -0,0 +1,304 @@
+            })
+            .partial()
+        ),
+        import: z.object({ importSchema }),
```
why is import schema specially included here?

FuzzB0t replied:
This is meant to be the importSpec returned by the import action

dbeal-eth commented on packages/builder/src/schemas.zod.ts:
```diff
@@ -0,0 +1,304 @@
+      .deepPartial()
+  );
+
+export const runSchema = z
```
should be removed, the run schema is only in cli

dbeal-eth commented on packages/builder/src/steps/contract.ts:
```diff
@@ -108,6 +82,8 @@ export default {
   },
 
   configInject(ctx: ChainBuilderContextWithHelpers, config: Config) {
+    validateConfig('contract', config);
```
rather than verifying the config at the beginning here why not change the `ChainDefinition` class to automatically validate before calling config inject? otherwise, this is likely to be missed by plugin integrators.

dbeal-eth commented on packages/builder/src/steps/invoke.ts:
```diff
@@ -245,10 +193,12 @@ export default {
 
     if (config.target) {
-      config.target = config.target.map((v) => _.template(v)(ctx));
+      config.target = config.target.map((v) => _.template(v)(ctx)) as [string, ...string[]];
```
comment explaining the use of unusual type would be helpful here. is this supposed to be a `string[]` with min length of 1?

FuzzB0t replied:
Will do. It means exactly what you described. This is the type that zod returns when setting that validation.

dbeal-eth reviewed (changes requested):
almost all the `.string()` defined types in the file can be further validated in some way, either as an intterpolated value, or as the ultimate type they should be. see comments.

do we have CI to verify that the generated doc pages are up to date?

im suer this was an intentional design decision, but please include the reasoning for having the jekyll docs checked into the repo (especially when github is not hosting I assume?)

btw I know we talked earlier but doing the detailed review I realized somethings. most of my comments are *suggestions* and/oir not serious.

FuzzB0t commented:
> Contributor

Good point, noted. Will add further validation to the string types.

As for the CI for docs I added this after our call today, I changed a bit about how the docs structure works and added that as part of it. Will push it along with the rest of the requested changes.

Thanks for your input! I think all of these suggestions are reasonable. The detailed review is super helpful.

> almost all the `.string()` defined types in the file can be further validated in some way, either as an intterpolated value, or as the ultimate type they should be. see comments.
> 
> do we have CI to verify that the generated doc pages are up to date?
> 
> im suer this was an intentional design decision, but please include the reasoning for having the jekyll docs checked into the repo (especially when github is not hosting I assume?)
> 
> btw I know we talked earlier but doing the detailed review I realized somethings. most of my comments are _suggestions_ and/oir not serious.

FuzzB0t commented on packages/builder/README.md:
Misc change, will fix this.

FuzzB0t commented:
An important thing to note about merging this PR:

The cannon-plugin-router's schemas will have to be changed to zod schemas for it to work on this version.

FuzzB0t commented:
This PR is ready to merge, could use one more review but I've gone pretty far testing manually. Will create a new issue to add a test suite to test for more edge cases and refine validations even further.
